accepted/2021/package-validation/package-validation.md

```diff
@@ -0,0 +1,190 @@
+
+```C#
+#if NET7_0_OR_GREATER
+    public void DownloadLogs(Utf8String area)
+    {
+        DownloadLogs(area.AsSpan());
+    }
+#else
+    public void DownloadLogs(string area)
+    {
+        var utf8Bytes = Encoding.UTF8.GetBytes(area);
+        DownloadLogs(utf8Bytes);
+    }
+#endif
+
+    private void DownloadLogs(ReadOnlySpan<byte> areaUtf8)
+    {
+        // Do network call
+    }
+```
+
+When Finley builds the project it fails with the following error:
+
+> error SYSLIB1234: The method 'DownloadLogs(string)' exists for .NET Standard
+> 2.0 but does not exist for .NET 7.0. This prevents consumers who compiled
+> against .NET Standard 2.0 to run on .NET 7.0.
+
+Finley understands that they shouldn't exclude `DownloadArea(string)` but
+instead just provide an additional `DownloadArea(Utf8String)` method for .NET
+7.0 and changes the code accordingly:
+
+```C#
+#if NET7_0_OR_GREATER
+    public void DownloadLogs(Utf8String area)
+    {
+        // ...
+    }
+#endif
+
+    public void DownloadLogs(string area)
+    {
+        // ...
+    }
+```
+
+### Validation against previous version
+
+Skylar is tasked with adding support for a connection time out to their library.
+The `Connect` method looks like this right now:
+
+```C#
+public static Client Connect(string url)
+{
+    // ...
+}
+```
+
+Since a connection timeout is an advanced configuration setting they reckon they
+can just add an optional parameter:
+
+```C#
+public static Client Connect(string url, TimeSpan? timeout = default)
+{
+    // ...
+}
+```
+
+However, when they rebuild the package they are getting an error:
+
+> error SYSLIB1234: The method 'Connect(string)' exists in the previous version
+> of the NuGet package (AdventureWorks.Client 1.2.3) but no longer exists in the
+> current version being built (1.2.4). This is a breaking change.
+
+Skylar realizes that while this is not a source breaking change, it's a binary
+breaking change. They solve this problem by adding an overload instead:
+
+```C#
+public static Client Connect(string url)
+{
+    return Client(url, Timeout.InfiniteTimeSpan);
+}
+
+public static Client Connect(string url, TimeSpan timeout)
+{
+    // ...
+}
+```
+
+### Making an intentional breaking change
+
+Armani is working on version 0.7 of their library. While they don't want to make
+accidental breaking changes, they do want to make changes based on user feedback.
+
+In version 0.6 they had a `Connect()` method that accepted a timeout as an
+`int`. Users complaint that timeouts in .NET are nowadays generally modelled
+using `TimeSpan` instead. After making the necessary code change, rebuilding the
+library results in an error:
+
+> error SYSLIB1234: The method 'Connect(string, int)' exists in the previous
+> version of the NuGet package (Fabrikam.Client 0.6.3) but no longer exists in
+> the current version being built (0.7.0). This is a breaking change.
+
+Since Armani's library is still an 0.x, which, according to SemVer, doesn't
```

Comment:
Does the analyzer look at the version number? I assume not

Reply:
I assume this would be a configuration that a user would set in their project as this feature would be opt-in.

Reply:
Oh, perhaps it will not report breaking changes if they are between two packages with different major versions? If so, is there a way to force it to do so anyway (as we would for our own packages)?

Reply:
I think I would always report errors and then users can suppress messages for intentional breaks. I would be scared of having a mode that runs by default and swallows errors that might help the user avoid a breaking change and rather have them being intentional about suppressing that warning.

Reply:
That makes sense. I know that for Microsoft, we may increment the major version for a variety of reasons, such as marketing, not because it's time to break stuff.

Reply:
I thought about that. Unless we're missing something, I don't believe we need configuration regarding the version number, but we will need an opt-in for previous version validation and we need a way to configure feed(s).

* The tool knows which version is being built.
* For each feed, the tool can request the existing versions.
* The previous version is the one that immediately precedes the current version in a combined sorted list

Originally I thought about a setting for "ignore breaking changes across major version boundaries" but it seems even if you do cross a sem ver boundary you still want to make intentional breaking changes only. The  tool only knows which ones are intentional if you the developer suppressed them, so I don't believe we need a SemVer setting at all.